packages/sdk-router/src/sdk.ts

```diff
@@ -39,141 +53,396 @@ class SynapseSDK {
+        return {
+          ...routerQuery,
+          swapAdapter: routerQuery.swapAdapter ?? routerQuery.routerAdapter,
+        };
+      });
     }
 
-    // Find the best query (in practice, we could return them all)
-    let destInToken
-    let rawOriginQuery
-    let rawDestQuery
-    let maxAmountOut: BigNumber = BigNumber.from(0)
-    for (let i = 0; i < destQueries.length; i++) {
-      if (destQueries[i].minAmountOut.gt(maxAmountOut)) {
-        maxAmountOut = destQueries[i].minAmountOut
-        rawOriginQuery = originQueries[i]
-        rawDestQuery = destQueries[i]
-        destInToken = bridgeTokens[i].token
+    public findBestQuery(
+      destQueries: RawQuery[],
+      originQueries: RawQuery[],
+      bridgeTokens: { symbol: string; token: string }[]
+    ): [RawQuery, RawQuery, { symbol: string; token: string }] {
+      let maxAmountOut: BigNumber = BigNumber.from(0);
+      let bestDestQuery: RawQuery | null = null;
+      let bestOriginQuery: RawQuery | null = null;
+      let bestBridgeToken: { symbol: string; token: string } | null = null;
+
+      for (let i = 0; i < destQueries.length; i++) {
+        if (destQueries[i].minAmountOut.gt(maxAmountOut)) {
+          maxAmountOut = destQueries[i].minAmountOut;
+          bestDestQuery = destQueries[i];
+          bestOriginQuery = originQueries[i];
+          bestBridgeToken = bridgeTokens[i];
+        }
       }
-    }
 
-    if (!rawOriginQuery || !rawDestQuery) {
-      throw Error('No route found')
+      if (!bestDestQuery || !bestOriginQuery || !bestBridgeToken) {
+        throw new Error('No best queries found');
+      }
+
+      return [bestOriginQuery, bestDestQuery, bestBridgeToken];
     }
 
-    // Set default deadline
-    const originQuery = convertQuery(rawOriginQuery)
-    originQuery.deadline = deadline ?? TEN_MIN_DEADLINE
-    const destQuery = convertQuery(rawDestQuery)
-    destQuery.deadline = ONE_WEEK_DEADLINE
 
-    // Get fee data
-    let feeAmount
-    let feeConfig
+    public async finalizeQuote(
+      bestQuery: [RawQuery, RawQuery, { symbol: string; token: string }],
+      router: SynapseRouter | SynapseCCTPRouter,
+      deadline: BigNumber | undefined,
+      isCCTP: boolean = false
+    ): Promise<any> {
+      const [bestOriginQuery, bestDestQuery, bestBridgeToken] = bestQuery;
+
+      // Set default deadlines
+      const originQuery = convertQuery(bestOriginQuery);
+      originQuery.deadline = deadline ?? TEN_MIN_DEADLINE;
+      const destQuery = convertQuery(bestDestQuery);
+      destQuery.deadline = ONE_WEEK_DEADLINE;
+
+      let feeAmount;
+      let feeConfig;
+
+      // Get fee data
+      if (isCCTP) {
+        // Cast router to SynapseCCTPRouter
+        const cctpRouter = <SynapseCCTPRouter>router;
+
+        feeAmount = await cctpRouter.routerContract.calculateFeeAmount(
+          bestBridgeToken.token,
+          originQuery.minAmountOut,
+          //TODO: how to determine isSwap?
```

Comment:
Swap Adapter field in destination query defines where there will be a swap or not:
`isSwap = (destQuery.swapAdapter != address(0))`